posts/2018-08-16-well-typed-heap.lhs

```diff
@@ -0,0 +1,1100 @@
+have as many choices for our minimal element as there are trees in our heap.  We
+need some way to write down this number as a type.
+
+Since we have a tree for every _1_ in our binary number, we can define the
+number of trees as the [popcount] of the binary number.
+
+[popcount]: https://en.wikichip.org/wiki/population_count
+
+In a weird twist of fate, you can also pretend this stands for "the count of
+trees which we can _pop_", which is exactly what we will be using it for.
+
+> type family Popcount (b :: Binary) :: Nat where
+>     Popcount 'BEnd   = 'Zero
+>     Popcount ('B1 b) = 'Succ (Popcount b)
+>     Popcount ('B0 b) = Popcount b
+
+`Popcount` can be used to relate the non-zeroness of a natural number, and the
+non-zeroness of a binary number.
+
+> lemma3
+>     :: BNonZero b ~ 'True
+>     => SBin b
+>     -> NNonZero (Popcount b) :~: 'True
+> lemma3 (SB1 _) = QED
+> lemma3 (SB0 b) = case lemma3 b of QED -> QED
+
+In addition to caring about the `popcount` of a binary number, we are sometimes
+interested in its `width` (number of bits).  This is also easily captured in a
+type family:
+
+> type family Width (binary :: Binary) :: Nat where
+>     Width 'BEnd        = 'Zero
+>     Width ('B0 binary) = 'Succ (Width binary)
+>     Width ('B1 binary) = 'Succ (Width binary)
+
+That is a fair amount of type families so far.  To make things a bit more clear,
+here is an informal visual overview of all the type families we have defined,
+including `BDec` (binary decrement, defined further below).
+
+![Rectangles represent (lifted) kinds, arrows are type families](/images/draft-05.png)
+
+<div id="cuttree"></div>
+
+Lumberjack
+----------
+
+Now, popping the smallest element from the heap first involves cutting a single
+tree from the forest inside the heap.  Afterwards, we take the root of that tree
+and merge the children of the tree back together with the original heap.
+
+![Selecting a single tree](/images/draft-06.png)
+
+However, just selecting (and removing) a single tree turns out to be quite an
+endeavour on its own.  We define an auxiliary GADT which holds the tree, the
+remainder of the heap, and most importantly a lot of invariants.
+
+Feel free to scroll down to the datatype from here if you are willing to assume
+the specific constraint and types are there for a reason.
+
+The two first fields are simply evidence singletons that we carry about.  `o`
+stands for the same concept as in `Forest`, it means we are starting with an
+order of `o`.  `x` stands for the index of the tree that was selected.
+
+This means the tree that was selected has an order of `NAdd o x`, as we can see
+in the third field.  If the remainder of the heap is `Forest o b a`, its shape
+is denoted by `b` and we can reason about the shape of the original heap.
+
+The children of tree (`Tree (NAdd o x) a`) that was selected will convert to
+heap of shape `Ones x`.  If we add the root to that, we get `BInc (Ones x)`.
```

Comment:
I don't understand what's going on here. Why are we adding a root the the heap of shape `Ones x`?

Reply:
Ok I understand why, but it was not easy to understand.

Reply:
Maybe emphasize that that we are working backwards somehow?
Or emphasize that the heap of size `BInc (Ones x)` is the heap with singleton tree `Tree (NAdd o x) a`